test/e2e/cosi_helpers.go

```diff
@@ -0,0 +1,74 @@
+//go:build e2e
+
+// Copyright 2025 Nutanix. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/test/framework"
+
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/api/v1alpha1"
+	apivariables "github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/api/variables"
+)
+
+type WaitForCOSIControllerToBeReadyInWorkloadClusterInput struct {
+	COSI                 *apivariables.COSI
+	WorkloadCluster      *clusterv1.Cluster
+	ClusterProxy         framework.ClusterProxy
+	DeploymentIntervals  []interface{}
+	HelmReleaseIntervals []interface{}
+}
+
+func WaitForCOSIControllerToBeReadyInWorkloadCluster(
+	ctx context.Context,
+	input WaitForCOSIControllerToBeReadyInWorkloadClusterInput, //nolint:gocritic // This hugeParam is OK in tests.
+) {
+	if input.COSI == nil {
+		return
+	}
+
+	switch ptr.Deref(input.COSI.Strategy, "") {
+	case v1alpha1.AddonStrategyHelmAddon:
+		WaitForHelmReleaseProxyReadyForCluster(
+			ctx,
+			WaitForHelmReleaseProxyReadyForClusterInput{
+				GetLister:       input.ClusterProxy.GetClient(),
+				Cluster:         input.WorkloadCluster,
+				HelmReleaseName: "cosi-controller",
+			},
+			input.HelmReleaseIntervals...,
+		)
+	case "":
+		Fail("COSI strategy is not set")
+	default:
+		Fail(
+			fmt.Sprintf(
+				"Do not know how to wait for COSI using strategy %s to be ready",
+				*input.COSI.Strategy,
+			),
+		)
+	}
+
+	workloadClusterClient := input.ClusterProxy.GetWorkloadCluster(
+		ctx, input.WorkloadCluster.Namespace, input.WorkloadCluster.Name,
+	).GetClient()
+
+	WaitForDeploymentsAvailable(ctx, framework.WaitForDeploymentsAvailableInput{
```

Comment:
nit:
This additional check is good. 
Does not helm Release proxy's Status will be in "Failed" state if the deployment fails?
I traced the code to https://github.com/helm/helm/blob/main/pkg/kube/wait.go#L63 where it checks for `Ready` before setting on the `status` field on the HelmReleaseProxy.

context: 
I am also looking into ways to report status of addon upgrade back to the CAREN clients.

Reply:
Good find! I copied this from existing code https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/blob/162915eb4794e176422c0d8d13c4bc6c2a57cc9b/test/e2e/csi_helpers.go#L316-L324

If its not necessary we can remove all of them in a future PR, WDYT?

Reply:
I don't think we wait for HCPs right now so as to not block the CAAPH control loop.

Reply:
we have this code to wait for HCP to be ready. https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/blob/main/pkg/handlers/generic/lifecycle/addons/helmaddon.go#L220C6-L220C19

It is only set to be invoked for the metallb addon. 
 > HCPs right now so as to not block the CAAPH control loop.
Is it because we dont want to wait/timeout returning from `AfterControlPlaneInitialization`/`BeforeClusterUpgrade` hook?

So cluster upgrade will continue even if a critical addon deployment failed?

Reply:
I am wondering if we should wait for both HCP and HRP to be ready for `upgrade addons` usecase.